File: Python/fieldOr.py
# Name: fieldOr.py
# Adds a new field to the attribute table which contains ones for features that are either in input field one or input field two true. Otherwise it will contain zero values. 
# Project: SimUSys, Christian Mueller, christian1.mueller@hs-bochum.de, 27.10.2016
# --------------------------------------------------------------------

import arcpy
import os
import subprocess
import sys


# inputs from ArcGIS
inShp = arcpy.GetParameterAsText(0)
inField1 = arcpy.GetParameterAsText(1)
inField2 = arcpy.GetParameterAsText(2)
outField = arcpy.GetParameterAsText(3)
pyScript = sys.argv[0]

# inShp = "C:/HochschuleBochum/Daten/Bochum/OSMData/Geostat_roads_workingData.shp"
# inField1 = "bridge_int"
# inField2 = "tunnel_int"
# outField = "brOTu"
# pyScript = "C:/HochschuleBochum/CodesScripts/Python/fieldOr.py"

allreadyPresent = False
allFields = arcpy.ListFields(inShp)
for i in range(0,len(allFields)):
    if allFields[i].name == outField:
        allreadyPresent = True


if (allreadyPresent == False):

    # execute R-Script
    thisScriptPath = os.path.dirname(pyScript)
    rScriptPath = os.path.join(os.path.abspath(os.path.join(thisScriptPath, os.pardir)), "R")
    rScript = os.path.join(rScriptPath, "fieldOr.r")
    arcpy.SetProgressor("default", "Executing R Script...")
    args = ["R", "--slave", "--vanilla", "--args", inShp, inField1, inField2, outField, rScriptPath] 
    scriptSource = open(rScript, 'r')
    rCommand = subprocess.Popen(args, stdin = scriptSource, stdout = subprocess.PIPE, stderr = subprocess.PIPE, shell = True)
    outString, errString = rCommand.communicate()
    scriptSource.close()
    if errString and "...completed execution of R-script" not in outString:
        arcpy.AddMessage(errString)

# send parameter to ArcGIS
arcpy.SetParameterAsText(4, outField)

File: R/fieldOr.R
# Christian Mueller, christian1.mueller@hs-bochum.de, Bochum University of Applied Sciences

# input from python
# inShp = "C:/HochschuleBochum/Daten/Bochum/OSMData/OSMData_roads.shp"
# inShp = "C:/HochschuleBochum/Daten/Bochum/Netzwerke/Fusswegenetz_korrigiert_workingData.shp"
# inField1 = "Tunnel_int"
# inField2 = "Bruecke_int"
# outField = "brOTu"
# wd  = 'C:\\HochschuleBochum\\CodesScripts\\R'


args <- commandArgs()
inShp = args[5]
inField1 = args[6]
inField2 = args[7]
outField  = args[8]
wd  = args[9]


# load packages
print("Loading R-Packages...")
options(repos = c(CRAN = "http://cran.rstudio.com"))
.libPaths(paste(wd, "libraries", sep = "/"))
for (i in 1:2){
  if (!require("foreign", lib.loc = paste(wd, "libraries", sep = "/"))) install.packages("foreign", lib = paste(wd, "libraries", sep = "/"))
}


# read attribute table
attTab <- read.dbf(paste(strsplit(inShp, ".", fixed = T)[[1]][1], "dbf", sep = "."), as.is = T)

# get field data
tunhit <- grep(x = colnames(attTab), pattern = substr(inField1, 1, 9))
dat1 <- attTab[,tunhit]
brihit <- grep(x = colnames(attTab), pattern = substr(inField2, 1, 9))
dat2 <- attTab[,brihit]

# add field
outdat <- numeric(nrow(attTab))
for (i in 1:nrow(attTab)){
  if ((dat1[i] != 0) || (dat2[i] != 0)) outdat[i] <- 1
}
outTab <- cbind(attTab, outdat)
colnames(outTab)[ncol(outTab)] <- outField

# write .dbf to file
write.dbf(outTab, file = paste(strsplit(inShp, ".", fixed = T)[[1]][1], "dbf", sep = "."))

# report to ArcGIS
print ("...finished execution of R-Script.")
